feat: add C implementation for `math/base/special/minmax`

Neerajpathak07 opened the pull request:
Progresses #649,
Resolves #2106.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for `math/base/special/minmax`
 ```c
#include <stdint.h>

double min;
double max;

stdlib_base_minmax( 4.0, -5.0, &min, &max );
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649 
- resolves #2106 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: adding C implementation

Neerajpathak07 committed:
chore: updating manifest

Neerajpathak07 committed:
chore: updating examples

Neerajpathak07 committed:
chore: updating addon.c

Neerajpathak07 committed:
chore: updating addon logic

Neerajpathak07 committed:
chore: updating examples

Neerajpathak07 committed:
chore: updating addon

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/minmax/src/addon.c:
```diff
@@ -0,0 +1,129 @@
+* @param info   callback data
+* @return       Node-API value
+*/
+static napi_value addon( napi_env env, napi_callback_info info ) {
```
Have kept the addon logic as this for now and going through https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/napi to have an idea for a utility for `void` functions.
for this file I have refactored this [code-block](lib/node_modules/@stdlib/math/base/special/minmax/src/addon.c).

Neerajpathak07 replied:
Turns out we can convert it a very simplified fashion like:-
```C
static napi_value addon( napi_env env, napi_callback_info info ) {
        STDLIB_NAPI_ARGV( env, info, argv, argc, 3 );
	STDLIB_NAPI_ARGV_DOUBLE( env, x, argv, 0 );
	STDLIB_NAPI_ARGV_DOUBLE( env, y, argv, 1 );
	STDLIB_NAPI_ARGV_FLOAT64ARRAY( env, z, zlen, argv, 2 );
        
        double min;
	double max;
	stdlib_base_minmax( x, y, &min, &max );

	double *op = (double *)z;
	op[ 0 ] = min;
	op[ 1 ] = max;

	return NULL;
}
```
just searching for an utility that can accumulate `void` function.

Neerajpathak07 replied:
something similar was also done in [`rempio2`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/rempio2/src/addon.c)

gunjjoshi replied:
@Neerajpathak07 You can do something like this:

In `addon.c`, you can have:

```c
#include "stdlib/math/base/special/minmax.h"
#include "stdlib/napi/export.h"
#include "stdlib/napi/argv.h"
#include "stdlib/napi/argv_double.h"
#include "stdlib/napi/argv_float64array.h"
#include <node_api.h>

/**
* Receives JavaScript callback invocation data.
*
* @param env    environment under which the function is invoked
* @param info   callback data
* @return       Node-API value
*/
static napi_value addon( napi_env env, napi_callback_info info ) {
    STDLIB_NAPI_ARGV( env, info, argv, argc, 3 );
    STDLIB_NAPI_ARGV_DOUBLE( env, x, argv, 0 );
    STDLIB_NAPI_ARGV_DOUBLE( env, y, argv, 1 );
    STDLIB_NAPI_ARGV_FLOAT64ARRAY( env, out, outlen, argv, 2 );
    stdlib_base_minmax( x, y, &out[ 0 ], &out[ 1 ] );
    return NULL;
}

STDLIB_NAPI_MODULE_EXPORT_FCN( addon )
```

And, for `native.js`:

```js
function minmax( x, y ) {
	var out = new Float64Array( 2 );
	addon( x, y, out );
	return [ out[ 0 ], out[ 1 ] ];
}
```

Neerajpathak07 replied:
@gunjjoshi Bingo!!
I had drafted a similar logic earlier but was confused on how to go about the void function. But you just clarified it thank you soo much!!

Neerajpathak07 commented:
@gunjjoshi LMK your opinion on this!!

Neerajpathak07 committed:
chore: refactoring addon

Neerajpathak07 committed:
chore: CI errors

Neerajpathak07 commented:
Not sure though why only C examples CI tests are failing for `#include <node_api.h>` and not for any other files.

anandkaranubc commented:
/stdlib merge

anandkaranubc committed:
Merge branch 'develop' into add-C-minmax

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc commented:
/stdlib merge

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,65 @@
+* @param x      First value
+* @param y      Second value
+* @param min    destination to store minimum value
+* @param max    destination to store maximum value
```
```suggestion
* @param x      first number
* @param y      second number
* @param min    destination to store the minimum value
* @param max    destination to store the maximum value
```

anandkaranubc replied:
```suggestion
* @param x      first number
* @param y      second number
* @param min    destination to store the minimum value
* @param max    destination to store the maximum value
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,65 @@
+#include <stdlib.h>
+
+/**
+* Evaluates the min and max of two values.
```
```suggestion
* Evaluates the minimum and maximum values.
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,65 @@
+* @param max    destination to store maximum value
+*
+* @example
+* #include <stdint.h>
```
There are no integers in the example. This can be removed

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,65 @@
+* double min;
+* double max;
+* stdlib_base_minmax( 3.14, NaN );
+* // returns [ NaN, NaN ]
```
No return. Function type is void.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,65 @@
+#include "stdlib/math/base/special/minmax.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/assert/is_negative_zero.h"
+#include <stdlib.h>
```
This is not needed.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+// MAIN //
+
+/**
+* Returns the minimum and maximum values.
```
```suggestion
* Returns the minimum and maximum values and assigns results to a provided output array.
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+function minmax( x, y ) {
+	var out = new Float64Array( 2 );
+	addon( x, y, out );
+	return [ out[ 0 ], out[ 1 ] ];
```
```suggestion
	return out;
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/lib/native.js:
```diff
@@ -0,0 +1,58 @@
+*
+* @example
+* var v = minmax( 3.14, 4.2 );
+* // returns [ 3.14, 4.2 ]
```
```suggestion
* // returns <Float64Array>[ 3.14, 4.2 ]
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/manifest.json:
```diff
@@ -0,0 +1,86 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/napi/argv",
+        "@stdlib/napi/argv-double",
+        "@stdlib/napi/argv-float64array",
+        "@stdlib/napi/export",
```
These are not needed for examples/benchmark.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/test/test.native.js:
```diff
@@ -0,0 +1,129 @@
+
+	v = minmax( NaN, 3.14 );
+	t.strictEqual( isnan( v[ 0 ] ), true, 'returns NaN' );
+	t.strictEqual( isnan( v[ 1 ] ), true, 'returns NaN' );
```
```suggestion
	t.strictEqual( isnan( v[ 0 ] ), true, 'returns expected value' );
```

Applies here and below.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/include/stdlib/math/base/special/minmax.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the minimum and maximum values.
```
```suggestion
* Evaluates the minimum and maximum values.
```

The function doesn't return anything.

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
bench: fix makefile and redorder variables

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/examples/c/example.c:
```diff
@@ -0,0 +1,34 @@
+
+	for ( i = 0; i < 10; i++ ) {
+		stdlib_base_minmax( x[ i ], y[ i ], &min, &max );
+		printf( "x: %lf => min: %lf, y: %lf, minmax(x, y): %lf\n", x[ i ], y[ i ], min, max );
```
```suggestion
		printf( "x: %lf, y: %lf => min: %lf, max: %lf\n", x[ i ], y[ i ], min, max );
```

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,63 @@
+#include "stdlib/math/base/assert/is_negative_zero.h"
+
+/**
+* Evaluates the minimum and maximum values in a single pass.
```
For the reviewers: I've updated the description, but if "Evaluates the minimum and maximum values." is sufficiently clear on its own, I will revert the changes.

Planeshifter replied:
```suggestion
* Returns the minimum and maximum values in a single pass.
```

anandkaranubc committed:
Update native.js

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc committed:
Update test.native.js

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/minmax/src/main.c:
```diff
@@ -0,0 +1,63 @@
+		*max = 0.0 / 0.0; // NaN
+		return;
+	}
+	if ( x == y && x == 0 ) {
```
```suggestion
	if ( x == y && x == 0.0 ) {
```

anandkaranubc committed:
Update lib/node_modules/@stdlib/math/base/special/minmax/src/main.c

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/minmax/README.md:
```diff
@@ -119,6 +119,100 @@ for ( i = 0; i < 100; i++ ) {
+
+#### stdlib_base_minmax( x, y, &min, &max )
+
+Evaluates the minimum and maximum values in a single pass.
```
To match JS description:

```suggestion
Returns the minimum and maximum values in a single pass.
```

anandkaranubc replied:
@Planeshifter The reason I changed it to "evaluates" was because the return type is void in C.

Planeshifter replied:
@anandkaranubc Hmm, that makes sense. Let's keep "Evaluates".

Planeshifter reviewed (approved):
Looks great! Thanks @Neerajpathak07 and everyone else who worked on this PR!
